src/mmr.rs

```diff
@@ -411,6 +547,64 @@ fn calculate_peaks_hashes<'a, T: 'a + Clone, M: Merge<Item = T>, I: Iterator<Ite
     Ok(peaks_hashes)
 }
 
+// Simplified version of `calculate_peaks_hashes`.
+fn calculate_positions_of_proof_items_via_peaks_hashes(
+    mut leaves: Vec<u64>,
+    mmr_size: u64,
+    mut proof_items_count: usize,
+) -> Result<Vec<u64>> {
+    if leaves.iter().any(|pos| pos_height_in_tree(*pos) > 0) {
+        return Err(Error::NodeProofsNotSupported);
+    }
+
+    // special handle the only 1 leaf MMR
+    if mmr_size == 1 && leaves.len() == 1 && leaves[0] == 0 {
+        return Ok(Default::default());
+    }
+    // ensure leaves are sorted and unique
+    leaves.sort_by_key(|pos| *pos);
+    leaves.dedup_by(|a, b| a == b);
+    let peaks = get_peaks(mmr_size);
+
+    let mut peaks_hashes: Vec<u64> = Vec::with_capacity(peaks.len() + 1);
+    for peak_pos in peaks {
+        let mut leaves: Vec<_> = take_while_vec(&mut leaves, |pos| *pos <= peak_pos);
+        if leaves.len() == 1 && leaves[0] == peak_pos {
+            // leaf is the peak
+            let _ = leaves.remove(0);
+        } else if leaves.is_empty() {
+            // if empty, means the next proof is a peak root or rhs bagged root
+            if proof_items_count > 0 {
+                proof_items_count -= 1;
+                peaks_hashes.push(peak_pos);
+            } else {
+                // means that either all right peaks are bagged, or proof is corrupted
+                // so we break loop and check no items left
+                break;
+            }
+        } else {
+            calculate_positions_of_proof_items_via_peak_root(
+                leaves,
+                peak_pos,
+                &mut proof_items_count,
+                &mut peaks_hashes,
+            )?;
+        };
+    }
+
+    // ensure nothing left in leaves
+    if !leaves.is_empty() {
+        return Err(Error::CorruptedProof);
+    }
+
+    // check rhs peaks
+    if proof_items_count > 0 {
+        // With lots of fuzz test, this condition is never entered.
+        unreachable!();
+    }
```

Comment:
Ref: #34.

Reply:
I think it's reachable, a malicious attacker can create a proof with duplicate items, please checkout https://github.com/nervosnetwork/merkle-mountain-range/pull/36 and add a similar to cover this case.

Reply:
I think so.